prover/crates/bin/witness_generator/src/basic_circuits.rs

```diff
@@ -446,43 +461,103 @@ async fn generate_witness(
+            },
         );
         (scheduler_witness, block_witness)
     })
     .instrument(make_circuits_span);
 
-    let mut circuit_urls = vec![];
-    let mut recursion_urls = vec![];
+    let mut save_circuit_handles = vec![];
 
-    let mut circuits_present = HashSet::<u8>::new();
+    // Ordering determines how we compose the circuit proofs in Leaf Aggregation Round.
+    // Sequence is used to determine circuit ordering (the sequencing of instructions) .
+    // If the order is tampered with, proving will fail (as the proof would be computed for a different sequence of instruction).
+    let circuit_sequence = AtomicUsize::new(0);
 
     let save_circuits_span = tracing::info_span!("save_circuits");
-    let save_circuits = async {
-        loop {
-            tokio::select! {
-                Some(circuit) = circuit_receiver.recv().instrument(tracing::info_span!("wait_for_circuit")) => {
-                    circuits_present.insert(circuit.numeric_circuit_type());
-                    circuit_urls.push(
-                        save_circuit(block_number, circuit, circuit_urls.len(), object_store).await,
-                    );
-                }
-                Some((circuit_id, queue, inputs)) = queue_receiver.recv().instrument(tracing::info_span!("wait_for_queue")) => {
-                    let urls = save_recursion_queue(block_number, circuit_id, queue, &inputs, object_store).await;
-                    recursion_urls.push(urls);
-                }
-                else => break,
-            };
+
+    // Future which receives circuits and saves them async.
+    let circuit_receiver_handle = async {
+        let semaphore = Arc::new(Semaphore::new(MAX_IN_FLIGHT_CIRCUITS));
+
+        while let Some(circuit) = circuit_receiver
+            .recv()
+            .instrument(tracing::info_span!("wait_for_circuit"))
+            .await
+        {
+            let sequence = circuit_sequence.fetch_add(1, Ordering::SeqCst);
+            let object_store = object_store.clone();
+            let semaphore = semaphore.clone();
+            save_circuit_handles.push(tokio::task::spawn(async move {
+                let _permit = semaphore
+                    .acquire()
+                    .await
+                    .expect("failed to get permit for running save circuit task");
+                save_circuit(block_number, circuit, sequence, object_store).await
+            }));
+        }
+    }
+    .instrument(save_circuits_span);
+
+    let mut save_queue_handles = vec![];
+
+    let save_queues_span = tracing::info_span!("save_queues");
+
+    // Future which receives recursion queues and saves them async.
+    let queue_receiver_handle = async {
+        while let Some((circuit_id, queue, inputs)) = queue_receiver
+            .recv()
+            .instrument(tracing::info_span!("wait_for_queue"))
+            .await
+        {
+            let object_store = object_store.clone();
```

Comment:
It was a genuine question, because from the code it looks like it was missed by accident. Probably worth a comment stating that we don't need semaphore here, because these jobs are small and there is not a lot of them.